### DISCUSSION TITLE
Walking `cue.Value` and Optional Fields

### DISCUSSION THREAD

**Discussion by nalum** [cue-lang/cue#2783] [Q&A] [answered]
I'm looking to use the golang library to implement a function that walks the structure and grabs all fields regardless of the field constraints, as in `first!:`, `second:` and `third?:`, so far I don't have issues with `first` and `second` but I've not been able to get `third` unless it's been given a value.

You can see the specific code I'm using here:

https://github.com/stefanprodan/timoni/blob/cfea950ba1f50195c6734777c2558366529baf32/internal/engine/module_builder.go#L312-L366

The goal is to build up a Markdown table that represents the configuration of a timoni module based on the actual configuration of that module and not have the module author maintain both the cue code and the Markdown table.

Any help is appreciated :bow:

**Comment by myitcv** [comment, marked as answer]
@Nalum thanks for raising. Take a look at this example:

```
go mod tidy
go run .
cmp stdout stdout.golden

-- go.mod --
module mod.com

go 1.21.4

require cuelang.org/go v0.7.0
-- main.go --
package main

import (
	"fmt"
	"log"

	"cuelang.org/go/cue"
	"cuelang.org/go/cue/cuecontext"
)

const src = `
x?: string
y!: string
z: string
`

func main() {
	ctx := cuecontext.New()
	v := ctx.CompileString(src)
	fields, err := v.Fields(
		cue.Optional(true),
		cue.Hidden(true),
		cue.Definitions(true),
	)
	if err != nil {
		log.Fatal(err)
	}
	for fields.Next() {
		fmt.Printf("found field %v\n", fields.Selector())
	}
}
-- stdout.golden --
found field x?
found field y!
found field z
```

We will also be adding more Go API-related howtos and tutorials to https://alpha.cuelang.org in the near future.

**Reply by nalum** [reply]
Thank you very much @myitcv. Working on the updated code here https://github.com/stefanprodan/timoni/pull/334

**Reply by jpluscplusm** [reply]
Doc tracked as https://github.com/cue-lang/docs-and-content/issues/112
